File: python/src/pmsec/tools/bun.py
from __future__ import annotations

from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import bun_config_path
from pmsec.util.version import build_preflight

NAME = "bun"
KEY = "minimumReleaseAge"
SECTION = "install"
DOCS = "https://bun.com/docs/runtime/bunfig#install"
MIN_BIN = (1, 3, 0)
EXTRAS: list[dict] = []

preflight = build_preflight(
    NAME, MIN_BIN,
    "minimumReleaseAge is silently ignored. Upgrade bun to enforce the cooldown.",
)


def path(ctx: Context) -> Path:
    return bun_config_path(ctx.env, ctx.home)


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY, section=SECTION)
    seconds = None
    if value is not None:
        try:
            seconds = int(value)
        except ValueError:
            seconds = None
    days = None if seconds is None else seconds // 86400
    return {"path": str(p), "configured": value, "days": days, "extras": []}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    write_atomic(p, set_key(raw, KEY, f"{KEY} = {days * 86400}", section=SECTION))
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed = remove_key(before, KEY, section=SECTION)
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}

File: node/src/tools/bun.mjs
import { bunConfigPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";
import { readExtras, applyExtras, removeExtras } from "../util/extras.mjs";
import { buildPreflight } from "../util/version.mjs";

export const name = "bun";
export const key = "minimumReleaseAge";
export const section = "install";
export const docs = "https://bun.com/docs/runtime/bunfig#install";
export const minBin = [1, 3, 0];
export const extras = [
  { key: "ignoreScripts", expected: "true", line: "ignoreScripts = true", section },
];

export function path(ctx) { return bunConfigPath(ctx.env, ctx.home); }

export const preflight = buildPreflight(name, minBin,
  "minimumReleaseAge is silently ignored. Upgrade bun to enforce the cooldown.");

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key, { section });
  const seconds = value === null ? null : Number(value);
  return {
    path: p, configured: value,
    days: seconds === null || Number.isNaN(seconds) ? null : Math.floor(seconds / 86400),
    extras: readExtras(raw, extras),
  };
}

export async function write(days, ctx) {
  const p = path(ctx);
  let text = setKey(await readSafe(p), key, `${key} = ${days * 86400}`, { section });
  text = applyExtras(text, extras);
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const cooldown = removeKey(before, key, { section });
  const ex = removeExtras(cooldown.text, extras);
  const removed = cooldown.removed || ex.removed;
  if (removed) await writeAtomic(p, ex.text);
  return { path: p, removed };
}
